File: module 1/week 1/examples/mergesort.go
package main

import (
	"fmt"
)

// MergeSorter is a structure for sorting arrays using Merge Sort.
type MergeSorter struct {
	array []int
}

// NewMergeSorter initializes a new instance of MergeSorter with the provided array.
func NewMergeSorter(array []int) *MergeSorter {
	if array == nil {
		panic("Input array cannot be nil")
	}
	return &MergeSorter{array: array}
}

// Sort performs the Merge Sort on the array without modifying the original array.
func (ms *MergeSorter) Sort() []int {
	if len(ms.array) <= 1 {
		return ms.array
	}
	return ms.mergeSort(ms.array)
}

// mergeSort recursively splits and sorts the array.
func (ms *MergeSorter) mergeSort(arr []int) []int {
	// Base case: if the array has 1 or 0 elements, it is already sorted.
	if len(arr) <= 1 {
		return arr
	}

	// Split the array into two halves
	mid := len(arr) / 2
	left := ms.mergeSort(arr[:mid])
	right := ms.mergeSort(arr[mid:])

	// Merge the sorted halves
	return ms.merge(left, right)
}

// merge combines two sorted slices into a single sorted slice.
func (ms *MergeSorter) merge(left, right []int) []int {
	merged := make([]int, 0, len(left)+len(right))
	i, j := 0, 0

	// Compare elements from both halves and add the smaller one to merged slice.
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			merged = append(merged, left[i])
			i++
		} else {
			merged = append(merged, right[j])
			j++
		}
	}

	// Append remaining elements from left or right slice.
	if i < len(left) {
		merged = append(merged, left[i:]...)
	} else {
		merged = append(merged, right[j:]...)
	}

	return merged
}

func main() {
	arr := []int{38, 27, 43, 3, 9, 82, 10}
	mergeSorter := NewMergeSorter(arr)
	sortedArr := mergeSorter.Sort()
	fmt.Println("Sorted array:", sortedArr)
}

File: course_1/week_1/examples/mergesort.py
class MergeSort:
    def __init__(self, array):
        if not isinstance(array, list):
            raise TypeError("Input should be a list.")
        self.array = array

    def sort(self):
        """Public method to start the merge sort."""
        if len(self.array) <= 1:
            return self.array
        return self._merge_sort(self.array)

    @staticmethod
    def _merge_sort(sub_array):
        """Recursively splits and sorts the array."""
        # Base case: if the list has 1 or 0 elements, it's already sorted.
        if len(sub_array) <= 1:
            return sub_array

        # Split the list into two halves.
        mid = len(sub_array) // 2
        left_half = MergeSort._merge_sort(sub_array[:mid])
        right_half = MergeSort._merge_sort(sub_array[mid:])

        # Merge the sorted halves.
        return MergeSort._merge(left_half, right_half)

    @staticmethod
    def _merge(left, right):
        """Merges two sorted halves into a single sorted list."""
        merged_array = []
        i = j = 0

        # Merge the two halves by comparing elements.
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                merged_array.append(left[i])
                i += 1
            else:
                merged_array.append(right[j])
                j += 1

        # Append any remaining elements from both halves.
        merged_array.extend(left[i:])
        merged_array.extend(right[j:])

        return merged_array

# Example usage.
arr = [38, 27, 43, 3, 9, 82, 10]
merge_sort_instance = MergeSort(arr)
sorted_arr = merge_sort_instance.sort()
print("Sorted array:", sorted_arr)
